File: java-crud/src/main/java/com/example/javacrud/service/ProductService.java
package com.example.javacrud.service;

import com.example.javacrud.model.Product;
import com.example.javacrud.repository.ProductRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProductRepository productRepository;

    public List<Product> findAll() {
        return productRepository.findAll();
    }

    public Product findById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Produto não encontrado com ID: " + id));
    }

    public Product save(Product product) {
        return productRepository.save(product);
    }

    public Product update(Long id, Product product) {
        Product existingProduct = findById(id);
        product.setId(existingProduct.getId());
        return productRepository.save(product);
    }

    public void delete(Long id) {
        Product product = findById(id);
        productRepository.delete(product);
    }
} 

File: kotlin-crud/src/main/kotlin/com/example/kotlincrud/service/ProductService.kt
package com.example.kotlincrud.service

import com.example.kotlincrud.model.Product
import com.example.kotlincrud.repository.ProductRepository
import jakarta.persistence.EntityNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class ProductService(private val productRepository: ProductRepository) {

    fun findAll(): List<Product> = productRepository.findAll()

    fun findById(id: Long): Product = productRepository.findById(id)
        .orElseThrow { EntityNotFoundException("Produto não encontrado com ID: $id") }

    @Transactional
    fun save(product: Product): Product = productRepository.save(product)

    @Transactional
    fun update(id: Long, product: Product): Product {
        findById(id) // Verifica se existe
        return productRepository.save(product.copy(id = id))
    }

    @Transactional
    fun delete(id: Long) {
        val product = findById(id)
        productRepository.delete(product)
    }
} 